cmd/notify-mailer/main.go

```diff
@@ -96,6 +98,90 @@ func (m *mailer) run() error {
 	return nil
 }
 
+// Resolves each reg ID to the most up-to-date contact email.
+// The email addresses are returned deduplicated and sorted.
+func (m *mailer) resolveDestinations() ([]string, error) {
+	var contacts []contact
+	err := json.Unmarshal(m.destinations, &contacts)
+	if err != nil {
+		return nil, err
+	}
+
+	// If there is no endpoint specified, use the total # of destinations
+	if m.checkpoint.end == 0 || m.checkpoint.end > len(contacts) {
+		m.checkpoint.end = len(contacts)
+	}
+
+	// Do not allow a start larger than the # of destinations
+	if m.checkpoint.start > len(contacts) {
+		return nil, fmt.Errorf(
+			"interval start value (%d) is greater than number of destinations (%d)",
+			m.checkpoint.start,
+			len(contacts))
+	}
+
+	contactMap := make(map[string]struct{}, len(contacts))
+	for _, c := range contacts[m.checkpoint.start:m.checkpoint.end] {
+		// Get the email address for the reg ID
+		emails, err := emailsForReg(c.ID, m.dbMap)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, email := range emails {
+			if strings.TrimSpace(email) == "" {
+				continue
+			}
+			// Using the contactMap to deduplicate addresses
+			contactMap[email] = struct{}{}
+		}
+	}
+
+	var contactsList []string
+	// Convert the de-dupe'd map back to a slice, sort it
```

Comment:
This creates a surprising property where emails are not sent in the same order they were in the input file, which makes it for instance hard to recover from errors.

When we think about this in terms of how we will use it for notifying people about lists of domains (e.g. for CAA SERVFAIL or malformed CSRs), it's not clear how we would do this coalescing. I think maybe we want to eliminate the de-duplication and say that you may receive multiple emails if you've created multiple accounts.